buildenv/jenkins/jobs/infrastructure/Cleanup-Nodes

```diff
@@ -0,0 +1,238 @@
+
+                                    // Clean up defunct pipelines workspaces
+                                    dir("${buildWorkspace}/../") {
+                                        retry(3) {
+                                            if (nodeLabels.contains('sw.os.aix')) {
+                                                sh "rm -rf Build* Test* PullRequest*"
+                                                sh "ls -al"
+                                            } else {
+                                                sh "ls | grep -v ${JOB_NAME} | xargs rm -rf"
+                                            }
+                                            sh "sleep 1m"
+                                        }
+                                    }
+                                }
+                            }
+
+                            if (MODES.contains('sanitize')) {
+                                stage("${nodeName} - Sanitize slave") {
+                                    // list user procs
+                                    sh "hostname && ps -f -u ${env.USER} | grep -EIv ${JOB_NAME} | grep -Eiv 'slave.jar' | grep -Eiv 'remoting.jar'| grep -Eiv '${env.USER}@notty' | sort"
+                                    // kill procs and reconnect slave
+                                    sanitize_node(nodeName)
+                                }
+                            }
+                        }
+                    }
+                }
+
+                if (offlineSlaves) {
+                    println("Offline slaves: ${offlineSlaves.toString()}")
+                }
+
+                if (unavailableSlaves) {
+                    println("Unavailable slaves: ${unavailableSlaves.toString()}")
+                }
+
+                parallel jobs
+
+            } catch(e) {
+                //slackSend channel: '#jenkins', color: 'danger', message: "Failed: ${env.JOB_NAME} #${env.BUILD_NUMBER} (<${env.BUILD_URL}|Open>)"
+                //slackSend channel: '@AdamBrousseau', color: 'danger', message: "Failed: ${env.JOB_NAME} #${env.BUILD_NUMBER} (<${env.BUILD_URL}|Open>)"
+                throw e
+            } finally {
+                cleanWs()
+            }
+        }
+    }
+}
+
+/*
+* Kill all processes and reconnect a Jenkins node
+*/
+def sanitize_node(nodeName) {
+    def workingNode = jenkins.model.Jenkins.instance.getNode(nodeName)
+    def workingComputer = workingNode.toComputer()
+
+    workingComputer.setTemporarilyOffline(true, null)
+    try {
+        println("\t ${nodeName}: Kill all processes...")
+        if (workingNode.getLabelString().indexOf("sw.os.zos") != -1) {
+            sh "ps -f -u ${env.USER} | awk '{print \$2}' | xargs kill -s KILL"
+        } else if (workingNode.getLabelString().indexOf("sw.os.windows") != -1) {
+            sh "ps -f -u ${env.USER} | awk '{print \$2}' | xargs kill -9"
+        } else {
+            //unix
+            sh "kill -9 -1"
```

Comment:
Why are we killing all processes? Can we _not_ kill the slave.jar and avoid having to reconnect the slave?

Reply:
To avoid re-connection it will require to skip the slave.jar and its children. The purpose is to sanitize the machine, kill all hanging processes safely and ensure the node is connected (avoid manual reconnection).

Reply:
I fail to see the need to kill the slave though. Unless I'm being obtuse to the process. I thought we could just grep out what we don't want like you did above.

Reply:
This is also an optional feature, run it when need it on all or a subset of machines.

Reply:
Will do reboot for windows only to ensure all procs are properly terminated.